beginner_source/onnx/onnx_registry_tutorial.py

```diff
@@ -0,0 +1,451 @@
+    )
+
+def to_numpy(tensor):
+    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
+
+onnx_input = export_output.adapt_torch_inputs_to_onnx(input_gelu_x)
+onnxruntime_input = {k.name: to_numpy(v) for k, v in zip(ort_session.get_inputs(), onnx_input)}
+onnxruntime_outputs = ort_session.run(None, onnxruntime_input)
+
+torch_outputs = aten_gelu_model(input_gelu_x)
+torch_outputs = export_output.adapt_torch_outputs_to_onnx(torch_outputs)
+
+assert len(torch_outputs) == len(onnxruntime_outputs)
+for torch_output, onnxruntime_output in zip(torch_outputs, onnxruntime_outputs):
+    torch.testing.assert_close(torch_output, torch.tensor(onnxruntime_output))
+
+
+######################################################################
+# Unsupported PyTorch operators with no ONNX RUNTIME support
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# In this case, the operator is not supported by any frameworks, and we
+# would like to use it in ONNX graph. Therefore, we need to implement
+# the operator in three places:
+#
+# 1. PyTorch FX graph
+# 2. ONNX registry
+# 3. ONNX RUNTIME
+#
+# In the following example, we would like to use a custom operator
+# that takes one tensor input, and returns an input. The operator adds
+# the input to itself, and returns the rounded result.
+#
+# **Custom Ops Registration in PyTorch FX Graph (Beta)**
+#
+# Firstly, we need to implement the operator in PyTorch FX graph.
+# This can be done by using torch._custom_op.
+#
+
+# NOTE: This is a beta feature in PyTorch, and is subject to change.
+from torch._custom_op import impl as custom_op
+
+@custom_op.custom_op("mylibrary::addandround_op")
+def addandround_op(tensor_x: torch.Tensor) -> torch.Tensor:
+    ...
+
+@addandround_op.impl_abstract()
+def addandround_op_impl_abstract(tensor_x):
+    return torch.empty_like(tensor_x)
+
+@addandround_op.impl("cpu")
+def addandround_op_impl(tensor_x):
+    # add x to itself, and round the result
+    return torch.round(tensor_x + tensor_x)
+
+torch._dynamo.allow_in_graph(addandround_op)
+
+class CustomFoo(torch.nn.Module):
+    def forward(self, tensor_x):
+        return addandround_op(tensor_x)
+
+input_addandround_x = torch.randn(3)
+custom_addandround_model = CustomFoo()
+
+
+######################################################################
+# **Custom Ops Registration in ONNX Registry**
+#
+# For the step 2 and 3, we need to implement the operator in ONNX registry.
+# In this example, we will implement the operator in ONNX registry
+# with the namespace "test.customop" and operator name "CustomOpOne",
+# and "CustomOpTwo". These two ops are registered and built in
+# `cpu_ops.cc <https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/testdata/custom_op_library/cpu/cpu_ops.cc>`__.
+#
+# Please make sure you have implemented the custom operators in cpp, and built
+# ONNX RUNTIME with the custom op library
```

Comment:
Link a reference to how to build ort from source here

Reply:
Removed as it's duplicated to the last section of this example.